state-chain/custom-rpc/src/lib.rs

```diff
@@ -1969,6 +1970,80 @@ pub enum NotificationBehaviour {
 	New,
 }
 
+/// Number of pinned blocks before starting to unpin old blocks. Must be lower than
+/// `ChainHeadConfig` maximum pinned blocks across all connections (ChainHeadConfig default is 512).
+const MAX_RETAINED_PINNED_BLOCKS: usize = 64;
+
+/// A subscription garbage collector that keeps track of pinned block hashes as reported by
+/// `chainHead_v1_follow` and unpins old blocks when the number of tracked pinned hashes reaches its
+/// configured capacity. The chain head API pins blocks in memory but clients have to explicitly
+/// call `chainHead_v1_unpin` to release memory. If pinned blocks reach MAX_PINNED_BLOCKS a `stop`
+/// event is generated causing all subscriptions to be dropped.
+/// SubscriptionCleaner is per subscription since blocks are pinned only in the context of a
+/// specific subscription. The api states that If multiple chainHead_v1_follow subscriptions exist,
+/// then each (subscription, block) tuple must be unpinned individually.
+struct SubscriptionCleaner<B: BlockT, BE: Backend<B>, C> {
+	chain_head_client: Arc<RpcModule<ChainHead<BE, B, C>>>,
+	sub_id: String,
+	pinned_hashes: Arc<Mutex<VecDeque<Hash>>>,
+}
+
+impl<B: BlockT, BE: Backend<B>, C> Clone for SubscriptionCleaner<B, BE, C> {
+	fn clone(&self) -> Self {
+		Self {
+			chain_head_client: self.chain_head_client.clone(),
+			sub_id: self.sub_id.clone(),
+			pinned_hashes: self.pinned_hashes.clone(),
+		}
+	}
+}
+
+impl<B: BlockT, BE: Backend<B>, C> SubscriptionCleaner<B, BE, C> {
+	pub fn new(
+		chain_head_client: Arc<RpcModule<ChainHead<BE, B, C>>>,
+		sub_id: String,
+		capacity: usize,
+	) -> Self {
+		Self {
+			chain_head_client,
+			sub_id,
+			pinned_hashes: Arc::new(Mutex::new(VecDeque::with_capacity(capacity))),
+		}
+	}
+
+	pub async fn add(&self, new_hashes: &[Hash]) {
+		let mut pinned_blocks = self.pinned_hashes.lock().await;
+
+		for new_hash in new_hashes.iter() {
+			// don't add duplicates
+			if !pinned_blocks.contains(new_hash) {
+				// Unpin the oldest block if we are at capacity
+				if pinned_blocks.len() + 1 >= pinned_blocks.capacity() {
+					if let Some(old_hash) = pinned_blocks.pop_front() {
+						if let Err(e) = self
+							.chain_head_client
+							.call::<_, ()>(
+								"chainHead_v1_unpin",
+								[&self.sub_id, &format!("{:?}", old_hash)],
+							)
```

Comment:
There is a jsonrpsee macro `rpc_params!` that makes this nicer to write:

```suggestion
							.call::<_, ()>(
								"chainHead_v1_unpin",
								jsonrpsee::rpc_params![self.sub_id, old_hash],
							)
```